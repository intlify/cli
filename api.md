# @intlify/cli API References

## Table Of Contents

- [Function](#function)
  - [compile](#compile)
- [Enum](#enum)
  - [CompileErrorCodes](#compileerrorcodes)
- [Interface](#interface)
  - [CompileOptions](#compileoptions)

## Function

### compile

Compile i18n resources

**Signature:**
```typescript
export declare function compile(source: string, output: string, options?: CompileOptions): Promise<boolean>;
```

#### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| source | string | the i18n resource source path, you can use glob pattern |
| output | string | the compiled i18n resource output path |
| options | CompileOptions | [CompileOptions](#compileoptions) |

#### Returns

 `true` when all i18n resource successfuly compile, not `false`

#### Remarks

This functoin is **asyncronous** function. If you want to get about error details, use the handler of [CompileOptions](#compileoptions) and [CompileErrorCodes](#compileerrorcodes)


## Enum

### CompileErrorCodes

Compile Error Codes

**Signature:**
```typescript
export declare const enum CompileErrorCodes 
```

#### Members

| Member | Value| Description |
| --- | --- | --- |
| INTERNAL_COMPILE_ERROR | 3 | Internal compile error |
| INTERNAL_COMPILE_WARNING | 2 | Internal compile warning |
| NOT_SUPPORTED_FORMAT | 1 | Not supported format |

#### Remarks

The error codes of [compile](#compile) function


## Interface

### CompileOptions

Compile Options

**Signature:**
```typescript
export interface CompileOptions 
```

#### Remarks

This optioins is used at [compile](#compile) function


#### Methods


#### Properties

##### onCompile

Compile handler

**Signature:**
```typescript
onCompile?: (source: string, output: string) => void;
```

##### onError

Compile Error handler

**Signature:**
```typescript
onError?: (code: number, source: string, output: string, msg?: string) => void;
```



