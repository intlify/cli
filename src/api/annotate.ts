import createDebug from 'debug'
import { parse } from '@vue/compiler-sfc'
import {
  getSFCBlocks,
  getSFCContentInfo,
  getCustomBlockContenType,
  escape
} from '../utils'

import type { SFCBlock, CompilerError } from '@vue/compiler-sfc'

const debug = createDebug('@intlify/cli:api:annotate')

/**
 * Annotate Warning Codes
 *
 * @remarks
 * The warning codes of {@link annotate} function
 *
 * @public
 */
export const enum AnnotateWarningCodes {
  /**
   * Not supported type
   */
  NOT_SUPPORTED_TYPE = 1,
  /**
   * Lang mismatch block `src` and block content
   */
  LANG_MISMATCH_IN_SRC_AND_CONTENT,
  /**
   * Lang mismatch option and block content
   */
  LANG_MISMATCH_IN_OPTION_AND_CONTENT,
  /**
   * Lang mismatch `lang` and block content
   */
  LANG_MISMATCH_IN_ATTR_AND_CONTENT
}

/* eslint-disable @typescript-eslint/no-explicit-any */
/**
 * Annotate options of {@link annotate} function
 *
 * @public
 */
export type AnnotateOptions = {
  /**
   * The type of the block
   *
   * @remarks
   * Only `i18n` type is supported, if you don't specify it. If any other type is specified, the function will raise the {@link SFCAnnotateError}.
   *
   * default as `i18n`
   */
  type?: string
  /**
   * Whether to force annotations
   *
   * @remarks
   * Force annotation of the attribute values of `attrs` option to the block tag. Even if the actual `lang` of the block content in the `lang` attribute is different, it will be enforced if this flag is turned on.
   * default as `false`
   */
  force?: boolean
  /**
   * The Attributes to be annotated on the block tag
   *
   * @remarks
   * default as `{}`
   */
  attrs?: Record<string, any>
  /**
   * The warning handler
   *
   * @remarks
   * Notify warnings generated by the annotate process
   */
  onWarn?: (code: number, args: Record<string, any>, block: SFCBlock) => void
}
/* eslint-enable @typescript-eslint/no-explicit-any */

/**
 * Vue SFC compiler error
 *
 * @remarks
 * This is the error wrapping the error that occurred in Vue SFC compiler
 *
 * @public
 */
export interface SFCParseError extends SyntaxError {
  /**
   * The error that occurred in Vue SFC compiler
   */
  erorrs: CompilerError[]
  /**
   * The filepath of the source file
   */
  filepath: string
}

/**
 * Check wheather Vue SFC compiler error
 *
 * @param err - The error that occurred in Vue SFC compiler
 * @returns if the error is raised in Vue SFC compiler, return `true`, else `false`
 *
 * @public
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isSFCParserError(err: any): err is SFCParseError {
  return 'errors' in err && 'filepath' in err
}

/**
 * The Annocation error
 *
 * @public
 */
export class SFCAnnotateError extends Error {
  /**
   * The filepath of the target file at annotate processing
   */
  filepath: string

  /**
   * Constructor
   *
   * @param message - The error message
   * @param filepath - The filepath of the target file at annotate processing
   */
  constructor(message: string, filepath: string) {
    super(message)
    this.name = 'SFCAnnotateError'
    this.filepath = filepath
  }
}

/* eslint-disable */
const NOOP_WARN = (
  code: number,
  args: Record<string, any>,
  block: SFCBlock
) => {}
/* eslint-enable */

/**
 * Annoate the Vue SFC block
 *
 * @param source - The source code of the Vue SFC
 * @param filepath - The file path of the Vue SFC
 * @param options - The {@link options | AnnotateOptions} of the annotate function
 * @returns The annotated source code of the Vue SFC
 *
 * @public
 */
export function annotate(
  source: string,
  filepath: string,
  options: AnnotateOptions = {}
): string {
  options.type = options.type || 'i18n'
  options.force = options.force || false
  options.attrs = options.attrs || {}
  options.onWarn = options.onWarn || NOOP_WARN

  const { descriptor, errors } = parse(source)
  if (errors.length) {
    debug('parse error', errors)
    const error = new SyntaxError(String('SFC parse error')) as SFCParseError
    error.erorrs = errors as CompilerError[]
    error.filepath = filepath
    throw error
  }

  // NOTE: currently support i18n only!
  if (options.type !== 'i18n') {
    throw new SFCAnnotateError('Not supported error', filepath)
  }

  const original = descriptor.source
  let offset = 0
  let diffset = 0
  let contents = [] as string[]

  contents = getSFCBlocks(descriptor).reduce((contents, block) => {
    debug(`start: type=${block.type}, offset=${offset}, diffset=${diffset}`)
    if (block.type !== options.type) {
      contents = contents.concat(original.slice(offset, block.loc.end.offset))
      offset = block.loc.end.offset
      return contents
    }

    const { content, lang } = getSFCContentInfo(block, filepath)
    const contentType = getCustomBlockContenType(content)
    debug('content info', block.lang, lang, contentType)
    if (contentType === 'unknwon') {
      options.onWarn!(
        AnnotateWarningCodes.NOT_SUPPORTED_TYPE,
        {
          type: block.type,
          actual: contentType
        },
        block
      )
      contents = contents.concat(original.slice(offset, block.loc.end.offset))
      offset = block.loc.end.offset
      return contents
    }

    if (block.src) {
      if (lang !== contentType) {
        options.onWarn!(
          AnnotateWarningCodes.LANG_MISMATCH_IN_SRC_AND_CONTENT,
          {
            src: lang,
            content: contentType
          },
          block
        )
      }
      contents = contents.concat(original.slice(offset, block.loc.end.offset))
      offset = block.loc.end.offset
      return contents
    }

    if (block.lang == null) {
      let lang = contentType as string
      if (options.attrs!.lang && options.attrs!.lang !== contentType) {
        options.onWarn!(
          AnnotateWarningCodes.LANG_MISMATCH_IN_OPTION_AND_CONTENT,
          {
            lang: options.attrs!.lang,
            content: lang
          },
          block
        )
        if (!options.force) {
          contents = contents.concat(
            original.slice(offset, block.loc.end.offset)
          )
          offset = block.loc.end.offset
          return contents
        } else {
          lang = options.attrs!.lang
        }
      }

      // annotate block tag attributes
      const startLoc = block.loc.start
      debug(
        `${block.type} block loc.start: offset=${startLoc.offset}, column=${startLoc.column}`
      )
      const blockTag = buildSFCBlockTag(block)
      const tagStartOffset = startLoc.offset - blockTag.length
      debug(`current tag: ${blockTag}`)
      debug(`tag start offset: ${tagStartOffset}`)
      const annoatedBlockTag = buildSFCBlockTag({
        type: options.type,
        attrs: { ...options.attrs!, lang }
      })
      debug(
        `annotated tag: ${annoatedBlockTag} (length:${annoatedBlockTag.length})`
      )
      const blockContent = `${annoatedBlockTag}${content}`
      debug(`content: ${blockContent}`)
      contents = contents.concat([
        original.slice(offset, tagStartOffset),
        blockContent
      ])
      diffset += annoatedBlockTag.length - blockTag.length
      offset = block.loc.end.offset

      return contents
    } else {
      if (lang !== contentType) {
        options.onWarn!(
          AnnotateWarningCodes.LANG_MISMATCH_IN_ATTR_AND_CONTENT,
          {
            lang,
            content: contentType
          },
          block
        )
      }
      contents = contents.concat(original.slice(offset, block.loc.end.offset))
      offset = block.loc.end.offset
      return contents
    }
  }, contents)
  debug(`end: offset=${offset}, diffset=${diffset}`)

  contents = contents.concat(original.slice(offset, original.length))
  return contents.join('')
}

function buildSFCBlockTag(meta: {
  type: string
  attrs: Record<string, any> // eslint-disable-line @typescript-eslint/no-explicit-any
}): string {
  let tag = `<${meta.type}`
  for (const [key, value] of Object.entries(meta.attrs)) {
    if (value === true) {
      tag += ` ${key}`
    } else {
      tag += ` ${key}="${escape(meta.attrs[key])}"`
    }
  }
  tag += '>'
  return tag
}
