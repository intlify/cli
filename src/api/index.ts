export * from './utils'
export * from './compile'
export * from './annotate'
export * from './format'
