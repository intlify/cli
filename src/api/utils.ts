import type { CompilerError } from '@vue/compiler-sfc'

/**
 * Vue SFC compiler error
 *
 * @remarks
 * This is the error wrapping the error that occurred in Vue SFC compiler
 *
 * @public
 */
export interface SFCParseError extends SyntaxError {
  /**
   * The error that occurred in Vue SFC compiler
   */
  erorrs: CompilerError[]
  /**
   * The filepath of the source file
   */
  filepath: string
}

/**
 * Check wheather Vue SFC compiler error
 *
 * @param err - The error that occurred in Vue SFC compiler
 * @returns if the error is raised in Vue SFC compiler, return `true`, else `false`
 *
 * @public
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isSFCParserError(err: any): err is SFCParseError {
  return 'errors' in err && 'filepath' in err
}
