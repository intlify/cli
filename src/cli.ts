import * as yargs from 'yargs'
import { initI18n, t } from './i18n'
;(async () => {
  await initI18n()
  yargs
    .scriptName('intlify')
    .usage(t('Usage: $0 <command> [options]'))
    .commandDir('./commands')
    .demandCommand()
    .help()
    .version().argv

  process.on('uncaughtException', err => {
    console.error(`uncaught exception: ${err}\n`)
    process.exit(1)
  })

  process.on('unhandledRejection', (reason, p) => {
    console.error('unhandled rejection at:', p, 'reason:', reason)
    process.exit(1)
  })
})()
