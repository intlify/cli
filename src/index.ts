export * from './generator'
export { CompileErrorCodes, CompileOptions, compile } from './api'
