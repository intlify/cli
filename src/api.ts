import path from 'path'
import { promises as fs } from 'fs'
import { debug as Debug } from 'debug'
import { generateJSON, generateYAML } from './generator/index'
import { globAsync } from './utils'

const SUPPORTED_FORMAT = ['.json', '.json5', '.yaml', '.yml']
const debug = Debug('@intlify/cli:api')

/**
 * Compile Error Codes
 *
 * @remarks
 * The error codes of {@link compile} function
 *
 * @public
 */
export const enum CompileErrorCodes {
  /**
   * Not supported format
   */
  NOT_SUPPORTED_FORMAT = 1,
  /**
   * Internal compile warning
   */
  INTERNAL_COMPILE_WARNING,
  /**
   * Internal compile error
   */
  INTERNAL_COMPILE_ERROR
}

/**
 * Compile Options
 *
 * @remarks
 * This optioins is used at {@link compile} function
 *
 * @public
 */
export interface CompileOptions {
  /**
   * Compile Error handler
   */
  onError?: (code: number, source: string, output: string, msg?: string) => void
  /**
   * Compile handler
   */
  onCompile?: (source: string, output: string) => void
}

/**
 * Compile i18n resources
 *
 * @param source - the i18n resource source path, you can use glob pattern
 * @param output - the compiled i18n resource output path
 * @param options - {@link CompileOptions}
 *
 * @remarks
 * This functoin is **asyncronous** function. If you want to get about error details, use the handler of {@link CompileOptions} and {@link CompileErrorCodes}
 *
 * @returns `true` when all i18n resource successfuly compile, not `false`
 *
 * @public
 */
export async function compile(
  source: string,
  output: string,
  options: CompileOptions = {}
): Promise<boolean> {
  let ret = true

  const targets = await globAsync(source)
  debug('compile: targets', targets)

  for (const target of targets) {
    const parsed = path.parse(target)
    debug('parsed', parsed)
    if (!parsed.ext) {
      continue
    }
    const filename = `${parsed.name}.js`
    const generatePath = path.resolve(output, filename)
    if (!SUPPORTED_FORMAT.includes(parsed.ext)) {
      options.onError &&
        options.onError(
          CompileErrorCodes.NOT_SUPPORTED_FORMAT,
          target,
          generatePath
        )
      ret = false
      continue
    }
    const source = await fs.readFile(target, { encoding: 'utf-8' })
    const generate = /\.json?5/.test(parsed.ext) ? generateJSON : generateYAML
    let occuredError = false
    const { code } = generate(source, {
      type: 'plain',
      filename: target,
      env: 'production',
      onError: (msg: string): void => {
        occuredError = true
        options.onError &&
          options.onError(
            CompileErrorCodes.INTERNAL_COMPILE_ERROR,
            target,
            generatePath,
            msg
          )
        ret = false
      },
      onWarn: (msg: string): void => {
        options.onError &&
          options.onError(
            CompileErrorCodes.INTERNAL_COMPILE_WARNING,
            target,
            generatePath,
            msg
          )
        ret = false
      }
    })
    if (!occuredError) {
      await writeGenerateCode(output, filename, code)
      options.onCompile && options.onCompile(target, generatePath)
    }
  }

  return ret
}

export async function writeGenerateCode(
  target: string,
  filename: string,
  code: string
): Promise<string> {
  await fs.mkdir(target, { recursive: true })
  const generatePath = path.resolve(target, filename)
  await fs.writeFile(generatePath, code)
  return generatePath
}
