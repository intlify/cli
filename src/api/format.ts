import createDebug from 'debug'
import { format as pritterFormat } from 'prettier'
import { parse } from '@vue/compiler-sfc'
import { getSFCBlocks, getSFCContentInfo } from '../utils'
import { SFCParseError } from './utils'

import type { CompilerError } from '@vue/compiler-sfc'
import { Options } from 'prettier'

const debug = createDebug('@intlify/cli:api:format')

/**
 * Format options of {@link format} function
 *
 * @public
 */
export type FormatOptions = {
  /**
   * The prettier options
   *
   * @remarks
   * The options for formatting the content of `i18n` custom blocks with prettier
   * default as {@link DEFAULT_PRETTIER_OPTIONS}
   */
  prettier?: Options
}

/**
 * Fortmat lang fwnot found error
 *
 * @reamrks
 * The error that not specified `lang` attribute in `i18n` custom block
 *
 * @public
 */
export class FormatLangNotFoundError extends Error {
  /**
   * The filepath of the target file at formatting processing
   */
  filepath: string

  /**
   * Constructor
   *
   * @param message - The error message
   * @param filepath - The filepath of the target file at formatting processing
   */
  constructor(message: string, filepath: string) {
    super(message)
    this.name = 'FormatLangNotFoundError'
    this.filepath = filepath
  }
}

/**
 * The default prettier options for formatting the content of `i18n` custom blocks
 *
 * @public
 */
export const DEFAULT_PRETTIER_OPTIONS = {
  printWidth: 100,
  tabWidth: 2,
  jsonRecursiveSort: true,
  plugins: ['prettier-plugin-sort-json']
}

/**
 * Format the Vue SFC block
 *
 * @remarks
 * Currently, only i18n custom blocks supporting
 *
 * @param source - The source code of the Vue SFC
 * @param filepath - The file path of the Vue SFC
 * @param options - The {@link FormatOptions | options} of the format function
 * @returns The formatted source code of the Vue SFC
 *
 * @public
 */
export function format(
  source: string,
  filepath: string,
  options: FormatOptions = {
    prettier: {}
  }
): string {
  const prettierOptions = Object.assign(
    {},
    DEFAULT_PRETTIER_OPTIONS,
    options.prettier
  )

  const { descriptor, errors } = parse(source)
  if (errors.length) {
    debug('parse error', errors)
    const error = new SyntaxError(String('SFC parse error')) as SFCParseError
    error.erorrs = errors as CompilerError[]
    error.filepath = filepath
    throw error
  }

  const original = descriptor.source
  let offset = 0
  let contents = [] as string[]

  contents = getSFCBlocks(descriptor).reduce((contents, block) => {
    debug(`start: type=${block.type}, offset=${offset}`)
    if (block.type !== 'i18n') {
      contents = contents.concat(original.slice(offset, block.loc.end.offset))
      offset = block.loc.end.offset
      return contents
    }

    const { content, lang } = getSFCContentInfo(block, filepath)
    if (lang == null) {
      throw new FormatLangNotFoundError('`lang` attr not found', filepath)
    }

    const startLoc = block.loc.start
    debug(
      `${block.type} block loc.start: offset=${startLoc.offset}, column=${startLoc.column}`
    )
    const formatted = pritterFormat(
      content,
      Object.assign({}, prettierOptions, { parser: lang })
    )
    const blockContent = `\n${formatted}`
    debug(`content: ${blockContent}`)
    contents = contents.concat([
      original.slice(offset, startLoc.offset),
      blockContent
    ])
    offset = block.loc.end.offset

    return contents
  }, contents)

  contents = contents.concat(original.slice(offset, original.length))
  return contents.join('')
}
