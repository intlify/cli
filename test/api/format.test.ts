import { assert, expect, describe, it } from 'vitest'
import { promises as fs } from 'fs'
import { dirname, resolve } from 'pathe'
import { format, FormatLangNotFoundError } from '../../src/api'

const __dirname = dirname(new URL(import.meta.url).pathname)

describe('format', function () {
  it('basic', async function () {
    const filepath = resolve(
      __dirname,
      '../fixtures/components/BasicFormat.vue'
    )
    const source = await fs.readFile(filepath, 'utf8')
    const content = await format(source, filepath)
    expect(content).toMatchSnapshot()
  })

  it('nested', async function () {
    const filepath = resolve(__dirname, '../fixtures/components/HasNested.vue')
    const source = await fs.readFile(filepath, 'utf8')
    const content = await format(source, filepath)
    expect(content).toMatchSnapshot()
  })

  it('oneline', async function () {
    const filepath = resolve(__dirname, '../fixtures/components/Oneline.vue')
    const source = await fs.readFile(filepath, 'utf8')
    const content = await format(source, filepath)
    expect(content).toMatchSnapshot()
  })

  it('multi', async function () {
    const filepath = resolve(
      __dirname,
      '../fixtures/components/MultiFormat.vue'
    )
    const source = await fs.readFile(filepath, 'utf8')
    const content = await format(source, filepath)
    expect(content).toMatchSnapshot()
  })

  it('vue 2', async function () {
    const filepath = resolve(
      __dirname,
      '../fixtures/components/FunctionalLangFill.vue'
    )
    const source = await fs.readFile(filepath, 'utf8')
    const content = await format(source, filepath, { vue: 2 })
    expect(content).toMatchSnapshot()
  })

  it('FormatLangNotFoundError', async function () {
    const filepath = resolve(__dirname, '../fixtures/components/Basic.vue')
    const source = await fs.readFile(filepath, 'utf8')
    let err
    try {
      await format(source, filepath)
    } catch (e) {
      err = e
    }
    assert.instanceOf(err, FormatLangNotFoundError)
  })
})
