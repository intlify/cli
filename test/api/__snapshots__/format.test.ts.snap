// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`format > basic 1`] = `
"<template>
  <p>This is BasicFormat</p>
</template>

<script>
export default {
  name: 'BasicFormat'
}
</script>

<style scoped>
#app {
  color: red;
}
</style>

<i18n lang=\\"json\\">
[object Promise]</i18n>
"
`;

exports[`format > multi 1`] = `
"<template>
  <p>This is MultiFormat</p>
</template>

<i18n lang=\\"json\\">
[object Promise]</i18n>

<script>
export default {
  name: 'MultiFormat'
}
</script>

<i18n lang=\\"yaml\\">
[object Promise]</i18n>

<style scoped>
#app {
  color: red;
}
</style>

<i18n lang=\\"json5\\">
[object Promise]</i18n>
"
`;

exports[`format > nested 1`] = `
"<template>
  <p>This is Nested resources</p>
</template>

<script>
export default {
  name: 'App'
}
</script>

<i18n lang=\\"json\\" global>
[object Promise]</i18n>

<style scoped>
#app {
  color: red;
}
</style>
"
`;

exports[`format > oneline 1`] = `
"<template>
  <p>This is Oneline</p>
</template>

<script>
export default { name: 'App' }
</script>

<style scoped>
#app { color: red; }
</style>

<i18n lang=\\"json\\">
[object Promise]</i18n>"
`;

exports[`format > vue 2 1`] = `
"<template functional>
  <p>This is functional component</p>
</template>

<i18n lang=\\"json\\">
[object Promise]</i18n>

<i18n lang=\\"json\\">
[object Promise]</i18n>"
`;
